ansible/roles/vm_set/files/mux_simulator.py

```diff
@@ -116,370 +112,535 @@ def run_cmd(cmdline):
+            target_out_sides = [out_side for out_side in self.flows['upstream']['out_sides'] if out_side not in out_sides]
+        else:
+            target_out_sides = list(set(self.flows['upstream']['out_sides'] + out_sides))
+
+        # Based on current out_sides and target out_sides to determine what to do
+        if set(self.flows['upstream']['out_sides']) == set(target_out_sides):
+            app.logger.debug('target_out_sides same as current out_sides, no upstream flow change required')
+            return
+        operation = None
+        if len(target_out_sides) == 0:
+            operation = 'DEL-FLOW'   # Remove the upstream flow
+        else:
+            if len(self.flows['upstream']['out_sides']) == 0:
+                operation = 'ADD-FLOW'   # Need to add upstream flow
+            else:
+                operation = 'MOD-FLOW'   # Need to modify upstream flow
+
+        if operation == 'DEL-FLOW':
+            run_cmd(DEL_FLOW_CMD.format(
+                    self.bridge,
+                    self.ports[NIC]))
+            self.flows['upstream']['out_sides'] = []
+        elif operation == 'ADD-FLOW':
+            action_desc = ','.join(['{}:"{}"'.format(OUTPUT, self.ports[out_side]) for out_side in target_out_sides])
+            run_cmd(ADD_FLOW_CMD.format(
+                    self.bridge,
+                    self.ports[NIC],
+                    action_desc))
+            self.flows['upstream']['out_sides'] = target_out_sides
+        elif operation == 'MOD-FLOW':
+            action_desc = ','.join(['{}:"{}"'.format(OUTPUT, self.ports[out_side]) for out_side in target_out_sides])
+            run_cmd(MOD_FLOW_CMD.format(
+                    self.bridge,
+                    self.ports[NIC],
+                    action_desc))
+            self.flows['upstream']['out_sides'] = target_out_sides
+        self.debug('updated upstream flow, new_action={}, out_sides={}, flows={}'.format(new_action, out_sides, json.dumps(self.flows, indent=2)))
+
+    def update_flows(self, new_action, out_sides):
+        """
+        Apply new flow action for the sides specified in out_sides.
+
+        Item in out_sides could be any of: 'nic', 'upper_tor', 'lower_tor'.
+        """
+        try:
+            self.lock.acquire()
+            self.info('>>>>> calling update_flows, new_action={}, out_sides={}, current flow:\n{}'.format(new_action, out_sides, json.dumps(self.flows, indent=2)))
+            if NIC in out_sides:
+                self._update_downstream_flow(new_action)
+            tor_sides = [out_side for out_side in out_sides if out_side != NIC]
+            if len(tor_sides) > 0:
+                self._update_upstream_flow(new_action, tor_sides)
+            self.info('update_flows completed, current flows:\n{} <<<<<<'.format(json.dumps(self.flows, indent=2)))
+        finally:
+            self.lock.release()
+
+    def reset_flows(self):
+        """Restore all the flows for this mux bridge.
```

Comment:
```reset_flows``` will toggle the mux to the last active side (that is ```self.active_side```) in ```_update_downstream_flow```. How about make it more determinate? Like toggle to the upper tor?

Reply:
Yes, currently `reset_flows` just ensure that all the flows output to their out side ports. The active side is not toggled. Just keep using the last active side `self.active_side`. Your suggestion sounds good, I'll make the change.

Reply:
After a second thought, I think it would be better just to restore the flows to output and do not always set active side to upper tor.

This API may be used by tests testing drop. Possibly at some point, the testing script need to simply restore all the flows and do not expect active/standby flap. If always set active side to upper_tor in reset_flows, it looks like a side effect of reset_flows. Anyway, we have other APIs for setting active side, right?

Reply:
Make sense to me. Let's keep the current design for now. Thanks